Fix second-pass queue cleanup when workload no longer needs retry

skools-here opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want faster PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Fixes cleanup of the second-pass scheduling queue by ensuring workloads are removed once they no longer require a retry, while preserving the initial backoff behavior.

#### Which issue(s) this PR fixes:
Fixes #8357

#### Special notes for your reviewer:
Cleanup is gated on retry iteration to avoid prematurely canceling the initial backoff window.

#### Does this PR introduce a user-facing change?

```release-note
TAS: Fixed an issue where workloads could remain in the second-pass scheduling queue (used for integration
or TAS with ProvisioningRequests, and for TAS Node Hot Swap) even if they no longer require to be in the queue.
```

mbobrovskyi commented:
/ok-to-test

skools-here commented:
/retest-required

gabesaba reviewed (commented):
Thanks for the fix! I just have a small nitpick comment

/approve

skools-here commented:
Hey @gabesaba I fixed the changes and added release notes also kindly review!

skools-here commented on pkg/cache/queue/manager.go:
```diff
@@ -815,16 +815,22 @@ func (m *Manager) DeleteSecondPassWithoutLock(wl *kueue.Workload) {
 // QueueSecondPassIfNeeded queues for the second pass of scheduling with exponential
 // delay.
 func (m *Manager) QueueSecondPassIfNeeded(ctx context.Context, w *kueue.Workload, iteration int) bool {
+	log := ctrl.LoggerFrom(ctx)
```
Added the unit coverage in TestQueueSecondPassIfNeeded for the cleanup path when a workload no longer requires a retry after the first iteration. Thanks!

skools-here commented:
> > Hey @gabesaba I fixed the changes and added release notes also kindly review!
> 
> thank you! will wait for @mimowo's comment to be addressed.

I have completed what @mimowo said kindly take a look !

skools-here commented on pkg/cache/queue/manager_test.go:
```diff
@@ -1449,6 +1456,13 @@ func TestQueueSecondPassIfNeeded(t *testing.T) {
+					baseWorkloadNotNeedingSecondPass.DeepCopy(),
+					1,
+				)
+			}
```
Thanks for the feedback! I’ll refactor the tests to use a dedicated fixture that updates the workload from “needs second pass” to “no longer needs second pass” and re invokes QueueSecondPassIfNeeded. I’ll push an update shortly.

skools-here commented:
I have pushed the changes in test can you please take a look.

mimowo commented on pkg/cache/queue/manager.go:
```diff
@@ -833,6 +839,13 @@ func (m *Manager) queueSecondPass(ctx context.Context, w *kueue.Workload, iterat
+	if !m.secondPassQueue.prequeued.Has(key) {
+		m.secondPassQueue.deleteByKey(key)
+		return
+	}
```
Oh sorry for misleading you maybe, I think this new code:

```
		wInfo := workload.NewInfo(w, m.workloadInfoOptions...)
		wInfo.SecondPassIteration = 0
```
does not seem relevant, because the wInfo would be GCed quickly.


Maybe to clarify I was thinking this entire block can be removed, because `prequeued` is arealdy checked inside `if m.secondPassQueue.queue(wInfo) {`

skools-here replied:
Thanks for the clarification! That makes sense  since `queue()` already checks `prequeued`, the extra guard here is redundant. I’ll simplify `queueSecondPass` accordingly and rely on the queue logic, keeping the tests to validate the cleanup behavior.

mimowo commented:
Since we are approaching the release let me try to take care of the CPs here.

tenzen-y commented on pkg/cache/queue/manager.go:
```diff
@@ -815,16 +815,22 @@ func (m *Manager) DeleteSecondPassWithoutLock(wl *kueue.Workload) {
 // QueueSecondPassIfNeeded queues for the second pass of scheduling with exponential
 // delay.
 func (m *Manager) QueueSecondPassIfNeeded(ctx context.Context, w *kueue.Workload, iteration int) bool {
+	log := ctrl.LoggerFrom(ctx)
```
```suggestion
	log := ctrl.LoggerFrom(ctx)
	wlKey := workload.Key(w)
```

`wlKey` is the same during this function.

tenzen-y commented on pkg/cache/queue/manager.go:
```diff
@@ -815,16 +815,22 @@ func (m *Manager) DeleteSecondPassWithoutLock(wl *kueue.Workload) {
+	} else if iteration > 0 {
+		// Remove the workload from the second-pass queue only after at least one
+		// retry iteration, to avoid canceling the initial backoff window.
+		// See #8357.
```
```suggestion
		// See https://github.com/kubernetes-sigs/kueue/issues/8357.
```

Could you use a link for visibility?

skools-here commented:
@tenzen-y thanks for the review I am working on your suggestions.

mimowo commented:
Thank you @skools-here for taking those 👍

mimowo commented on pkg/cache/queue/manager_test.go:
```diff
@@ -1395,22 +1396,26 @@ func TestQueueSecondPassIfNeeded(t *testing.T) {
 	cases := map[string]struct {
 		workloads []*kueue.Workload
-		deleted   sets.Set[workload.Reference]
+		update    func(*kueue.Workload)
```
I opened the potentially related issue https://github.com/kubernetes-sigs/kueue/issues/8633

mimowo commented:
Ok, I opened https://github.com/kubernetes-sigs/kueue/issues/8682 to keep track of that
